src/pages/academy/sourcereel/Sourcereel.tsx

```diff
@@ -116,294 +113,322 @@ export type StateProps = {
-        handleDebuggerPause={this.props.handleDebuggerPause}
-        handleDebuggerReset={this.props.handleDebuggerReset}
-        handleDebuggerResume={this.props.handleDebuggerResume}
-        handleEditorEval={editorEvalHandler}
-        handleInterruptEval={this.props.handleInterruptEval}
-        handleToggleEditorAutorun={this.props.handleToggleEditorAutorun}
-        isDebugging={this.props.isDebugging}
-        isEditorAutorun={this.props.isEditorAutorun}
-        isRunning={this.props.isRunning}
-        key="autorun"
-      />
-    );
+  const getTimerDuration = () => props.timeElapsedBeforePause + Date.now() - props.timeResumed;
 
-    const chapterSelectHandler = ({ chapter }: { chapter: Chapter }, e: any) => {
-      this.props.handleChapterSelect(chapter);
-      if (this.props.recordingStatus !== RecordingStatus.recording) {
-        return;
-      }
-      this.props.handleRecordInput({
-        time: this.getTimerDuration(),
-        type: 'chapterSelect',
-        data: chapter
-      });
+  const handleRecordInit = () => {
+    const initData: PlaybackData['init'] = {
+      chapter: props.sourceChapter,
+      externalLibrary: props.externalLibraryName as ExternalLibraryName,
+      // TODO: Hardcoded to make use of the first editor tab. Rewrite after editor tabs are added.
+      editorValue: props.editorTabs[0].value
     };
+    dispatch(recordInit(initData, workspaceLocation));
+  };
 
-    const chapterSelect = (
-      <ControlBarChapterSelect
-        handleChapterSelect={chapterSelectHandler}
-        sourceChapter={this.props.sourceChapter}
-        sourceVariant={this.props.sourceVariant}
-        key="chapter"
-      />
-    );
+  const handleRecordPause = () =>
+    props.handleRecordInput({
+      time: getTimerDuration(),
+      type: 'forcePause',
+      data: null
+    });
 
-    const clearButton = (
-      <ControlBarClearButton
-        handleReplOutputClear={this.props.handleReplOutputClear}
-        key="clear_repl"
-      />
-    );
+  const editorEvalHandler = () => {
+    props.handleEditorEval();
+    if (props.recordingStatus !== RecordingStatus.recording) {
+      return;
+    }
+    props.handleRecordInput({
+      time: getTimerDuration(),
+      type: 'keyboardCommand',
+      data: KeyboardCommand.run
+    });
+  };
+  const autorunButtons = (
+    <ControlBarAutorunButtons
+      handleDebuggerPause={() => dispatch(beginDebuggerPause(workspaceLocation))}
+      handleDebuggerResume={() => dispatch(debuggerResume(workspaceLocation))}
+      handleDebuggerReset={() => dispatch(debuggerReset(workspaceLocation))}
+      handleEditorEval={editorEvalHandler}
+      handleInterruptEval={() => dispatch(beginInterruptExecution(workspaceLocation))}
+      handleToggleEditorAutorun={() => dispatch(toggleEditorAutorun(workspaceLocation))}
+      isDebugging={props.isDebugging}
+      isEditorAutorun={props.isEditorAutorun}
+      isRunning={props.isRunning}
+      key="autorun"
+    />
```

Comment:
I agree that there's a lot of code duplication. I actually have a local branch that implements your second approach, but since it relies on the refactor PRs (I squashed them out of order in my local branch) and I didn't want to keep force pushing to reflect the changes to master I haven't pushed it out yet.